core/src/main/scala/kafka/tools/StorageTool.scala

```diff
@@ -45,89 +45,124 @@ import scala.jdk.CollectionConverters._
-          val metadataRecords : ArrayBuffer[ApiMessageAndVersion] = ArrayBuffer()
-          val specifiedFeatures: util.List[String] = namespace.getList("feature")
-          val releaseVersionFlagSpecified = namespace.getString("release_version") != null
-          if (releaseVersionFlagSpecified && specifiedFeatures != null) {
-            throw new TerseFailure("Both --release-version and --feature were set. Only one of the two flags can be set.")
-          }
-          val featureNamesAndLevelsMap = featureNamesAndLevels(Option(specifiedFeatures).getOrElse(Collections.emptyList).asScala.toList)
-          val metadataVersion = getMetadataVersion(namespace, featureNamesAndLevelsMap,
-            Option(config.get.originals.get(ReplicationConfigs.INTER_BROKER_PROTOCOL_VERSION_CONFIG)).map(_.toString))
-          validateMetadataVersion(metadataVersion, config)
-          // Get all other features, validate, and create records for them
-          // Use latest default for features if --release-version is not specified
-          generateFeatureRecords(
-            metadataRecords,
-            metadataVersion,
-            featureNamesAndLevelsMap,
-            Features.PRODUCTION_FEATURES.asScala.toList,
-            config.get.unstableFeatureVersionsEnabled,
-            releaseVersionFlagSpecified
-          )
-          getUserScramCredentialRecords(namespace).foreach(userScramCredentialRecords => {
-            if (!metadataVersion.isScramSupported) {
-              throw new TerseFailure(s"SCRAM is only supported in metadata.version ${MetadataVersion.IBP_3_5_IV2} or later.")
-            }
-            for (record <- userScramCredentialRecords) {
-              metadataRecords.append(new ApiMessageAndVersion(record, 0.toShort))
-            }
-          })
-
-          val bootstrapMetadata = buildBootstrapMetadata(metadataVersion, Some(metadataRecords), "format command")
-          val ignoreFormatted = namespace.getBoolean("ignore_formatted")
-          if (!configToSelfManagedMode(config.get)) {
-            throw new TerseFailure("The kafka configuration file appears to be for " +
-              "a legacy cluster. Formatting is only supported for clusters in KRaft mode.")
-          }
-          Exit.exit(formatCommand(System.out, directories, metaProperties, bootstrapMetadata,
-                                  metadataVersion,ignoreFormatted))
+      exitCode = runMain(args)
+    } catch {
+      case e: TerseFailure =>
+        exitCode = 1
+        message = Some(e.getMessage)
+    }
+    message.foreach(System.err.println)
+    Exit.exit(exitCode, message)
+  }
 
-        case "random-uuid" =>
-          System.out.println(Uuid.randomUuid)
-          Exit.exit(0)
+  private def parseConfig(configFilename: String): KafkaConfig = {
+    try {
+      new KafkaConfig(Utils.loadProps(configFilename))
+    } catch {
+      case e: IllegalArgumentException => throw new TerseFailure(s"Invalid configuration: ${e.getMessage}")
```

Comment:
Wrapping `IllegalArgumentException` slightly changes the behavior in case of bad configurations. For example if I set `log.dirs=`, on trunk I get:
```
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: At least one log directory must be defined via log.dirs or log.dir.
	at scala.Predef$.require(Predef.scala:337)
	at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:2316)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:2285)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1634)
	at kafka.tools.StorageTool$.$anonfun$main$1(StorageTool.scala:52)
	at scala.Option.flatMap(Option.scala:283)
	at kafka.tools.StorageTool$.main(StorageTool.scala:52)
	at kafka.tools.StorageTool.main(StorageTool.scala)
```

With this PR I get:
```
Invalid configuration: requirement failed: At least one log directory must be defined via log.dirs or log.dir.
```

Reply:
I reverted this with the config parsing issue mentioned above.